[RSDK-515] Modular Resources

Otterverse opened the pull request:
Lots of changes involved, so I'm not expecting any speedy reviews here. Also, I'd like at least one solid pass of reviews (especially about the higher level architecture) before starting to create proper tests for the new module methods as well.

As for where to look, two main places to start with: There is a new "module" section in the top level, which includes a doc.go file with a fair bit of explanation for how this is expected to work. https://github.com/Otterverse/rdk/blob/modular-resources/module/doc.go

The most obvious entrypoint though, is a through set of examples collected in https://github.com/Otterverse/rdk/tree/modular-resources/examples/customresources There is a README there with instructions for how to run things, and it's all tied up in a bow with makefile targets, so it should "just run."

Other points of interest include:

- Modules section as top-level of config
- Changing the concept of "model" from a single string, to a namespaced triplet. Ex: acme:demo:mygizmo
- IsMoving() moved from Local[COMPONENT] to just [COMPONENT] since this must be checkable via grpc for modular stuff

Lots of other changes (obviously) as well. Please don't hesitate to ask me questions via Slack! This needs to all make sense to everyone, and be the right way forward.

Also, note that this depends on patches to both api and goutils. PRs are in place for those as well.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/viamrobotics/rdk?pullRequest=1363) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/viamrobotics/rdk?pullRequest=1363) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/viamrobotics/rdk?pullRequest=1363) it.</sub>

edaniels commented:
@Otterverse can you put up PRs for goutils/api so we can at least refer to those changes?

edaniels commented:
is triplet change backwards compatible? rather, if we merged this in right now, are all configs backwards compatible?

stevebriskin commented on components/gripper/server.go:
```diff
@@ -66,3 +66,16 @@ func (s *subtypeServer) Stop(ctx context.Context, req *pb.StopRequest) (*pb.Stop
 }
+
+// IsMoving queries of a component is in motion.
+func (s *subtypeServer) IsMoving(ctx context.Context, req *pb.IsMovingRequest) (*pb.IsMovingResponse, error) {
```
did we miss adding all the IsMoving methods before? @cheukt

cheukt replied:
they were all added. we originally made isMoving an explicitly local method but I don't think we can keep it as is under this scheme because then there is no way for the "parent" robot to query for whether the modular resource is moving or not.

Otterverse replied:
Correct, it was local-only (and thus not in the grpc server/client) and I moved it to be a full/normal method. I think it has to be like this, because it can be safety critical (so can't just skip/workaround), and everything modular is on the other side of a grpc call.

Otterverse commented on config/watcher.go:
```diff
@@ -132,26 +134,28 @@ func newFSWatcher(ctx context.Context, configPath string, logger golog.Logger) (
-						return
-					case configCh <- newConfig:
-					}
+					debounced(func() {
+						//nolint:gosec
+						rd, err := os.ReadFile(configPath)
+						if err != nil {
+							logger.Errorw("error reading config file after write", "error", err)
+							return
+						}
+						if bytes.Equal(rd, lastRd) {
+							return
+						}
+						lastRd = rd
+						newConfig, err := FromReader(cancelCtx, configPath, bytes.NewReader(rd), logger)
+						if err != nil {
+							logger.Errorw("error reading config after write", "error", err)
+							return
+						}
+						select {
+						case <-cancelCtx.Done():
+							return
+						case configCh <- newConfig:
+						}
+					})
```
This was a driveby for a bug. Previously the filewatcher would see my IDE save the file, but it'd trigger on the very first write operation, whereas it actually takes many writes to fully put the file to disk. When that happened, it'd read before the file was complete, and get EOF errors and broken configs. This makes sure it only tries to read after no writes are seen for 500ms.

edaniels replied:
Ah, got it

Otterverse commented on module/modmanager/modmaninterface/interface.go:
```diff
@@ -0,0 +1,27 @@
+type ModuleManager interface {
+	AddModule(ctx context.Context, cfg config.Module) error
+
+	AddComponent(ctx context.Context, cfg config.Component, deps []string) (interface{}, error)
```
Ideally, deps should/could be the registry.Dependencies here, so the conversion to strings is internal, but that causes an import cycle in lots of places, so taking the []string instead.

Otterverse commented on module/module.go:
```diff
@@ -0,0 +1,422 @@
+		handlers:   make(HandlerMap),
+		services:   make(map[resource.Subtype]subtype.Service),
+	}
+	m.buildHandlerMap()
```
This builds a handlermap dynamically from things in the registry. In other languages/SDKs, it may make more sense to just let the module author define the map statically.

Otterverse commented on module/module.go:
```diff
@@ -0,0 +1,422 @@
+}
+
+// GetParentComponent returns a component from the parent robot by name.
+func (m *Module) GetParentComponent(ctx context.Context, name resource.Name) (interface{}, error) {
```
Should probably be renamed GetParentResource().

Otterverse commented on module/module.go:
```diff
@@ -0,0 +1,422 @@
+}
+
+// AddComponent receives the component configuration from the parent.
+func (m *Module) AddComponent(ctx context.Context, req *pb.AddComponentRequest) (*pb.AddComponentResponse, error) {
```
This function also relies heavily on the registry concept and our internal subtypeService system. May need to be rethought for other SDKs. E.g. a simple module might simply have the module author set a callback and then handle this directly.

Otterverse commented on module/module.go:
```diff
@@ -0,0 +1,422 @@
+}
+
+// AddService receives the service configuration from the parent.
+func (m *Module) AddService(ctx context.Context, req *pb.AddServiceRequest) (*pb.AddServiceResponse, error) {
```
AddService is nearly identical to AddComponent. The only reason it's different is that config.Component and config.Service are nominally different types (though contain pretty much the same structure these days) and because the service constructor gets the entire parent robot, instead of dependencies. Once service deps are added in RDK, I think we should consider merging components/services as just "resources" and vastly simplify things, including here.

Otterverse commented on registry/registry.go:
```diff
@@ -36,8 +37,7 @@ type RegDebugInfo struct {
 type Service struct {
 	RegDebugInfo
-	Constructor           CreateService
-	AttributeMapConverter config.AttributeMapConverter
```
We register attribute map converters independently now, so this was just driveby cleanup.

Otterverse commented on registry/registry.go:
```diff
@@ -254,13 +262,27 @@ func lookupSubtype(subtypeName resource.SubtypeName) (*resource.Subtype, bool) {
 	return validModels
 }
+
+// ReconfigurableComponent is implemented when component/service of a robot is reconfigurable.
+type ReconfigurableComponent interface {
+	// Reconfigure reconfigures the resource
+	Reconfigure(ctx context.Context, cfg config.Component, deps Dependencies) error
+}
+
+// ReconfigurableService is implemented when component/service of a robot is reconfigurable.
+type ReconfigurableService interface {
+	// Reconfigure reconfigures the resource
+	Reconfigure(ctx context.Context, cfg config.Service) error
+}
```
These were added here because they cause import cycles in the other places that also made sense (config and resource.)

Also, these are part of the overall workaround for RDK's current "reconfigure" system (which doesn't really reconfigure, but instead swaps out proxied objects.) I'll be writing up a blog/proposal to clean this all up within RDK itself in the not-too-distant future.

Otterverse commented on resource/model.go:
```diff
@@ -0,0 +1,140 @@
+	return fmt.Sprintf("%s:%s", m.ModelFamily, m.Name)
+}
+
+// UnmarshalJSON pareses namespace:family:modelname strings to the full Model{} struct.
+func (m *Model) UnmarshalJSON(data []byte) error {
```
While the new "Model" type is structured in a way just like Subtype is, this makes it easy to simply define the model in one human readable string instead of a nested object. I find it easier/simpler to read, and think we should default to that for examples/documentation.

Otterverse commented on robot/web/web.go:
```diff
@@ -220,9 +235,12 @@ type webService struct {
 	mu           sync.Mutex
 	r            robot.Robot
 	rpcServer    rpc.Server
+	liteServer   *googlegrpc.Server
```
I initially thought about trying to modify rpc.Server to support the unix socket stuff directly, so I could simply start two of those, and not maintain a separate concept, but... On the module side, I believe directly using google's Server keeps things easier to implement for third parties.

edaniels replied:
I think I agree. Just need to establish the security considerations around it.

Otterverse replied:
This has been reworked so that a new moduleServer object is used that fulfills rpc.Server, so no extra registration is required.

Otterverse commented on web/server/entrypoint.go:
```diff
@@ -307,7 +307,6 @@ func (s *robotServer) serveWeb(ctx context.Context, cfg *config.Config) (err err
 	defer func() {
 		<-onWatchDone
 	}()
-	defer cancel()
```
Drive-by... cancel() was already deferred way above.

Otterverse commented:
> is triplet change backwards compatible? rather, if we merged this in right now, are all configs backwards compatible?

It should be. UnmarshallJSON in resource/model.go should handle it. E.g. "xarm6" becomes "rdk:builtin:xarm6" effectively.

Otterverse commented:
> @Otterverse can you put up PRs for goutils/api so we can at least refer to those changes? Or just commit refs

Done!

Goutils: https://github.com/viamrobotics/goutils/pull/101

API: https://github.com/viamrobotics/api/pull/95

edaniels commented on module/modmanager/modmaninterface/interface.go:
```diff
@@ -0,0 +1,27 @@
+type ModuleManager interface {
+	AddModule(ctx context.Context, cfg config.Module) error
+
+	AddComponent(ctx context.Context, cfg config.Component, deps []string) (interface{}, error)
```
how close can we get to having this be just for resources from an API perspective. they're so close

Otterverse replied:
I know! It's so frustrating because the only difference (now that services have dependencies) is that config.Component is a different type than config.Service. I need to do the rebase/merge to bring that in, and then see what else would be needed.

Otterverse replied:
AddComponent/Service is combined now, using that hack-ish thing we discussed today.

edaniels commented on module/modmanager/manager.go:
```diff
@@ -0,0 +1,370 @@
+
+	conn, err := grpc.Dial(
+		"unix://"+mod.addr,
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
```
I'm worried about sniffing reads as well

Otterverse replied:
The module (where the socket is started) raises the umask before start, such that the socket itself should have 0600 permissions. E.g. only the owning user (or, of course, root) can read/write to it. No other normal user should have access at all. That said, I hadn't fully considered the race condition where someone CREATES a socket and tricks us into using it. Should be simple though. Putting comments over in module explaining that part.

Otterverse replied:
Resolving this and related comments per our discussion on slack. Will callout my improvements on this once I push my next commit.

edaniels commented on module/modmanager/manager.go:
```diff
@@ -0,0 +1,370 @@
+	for _, module := range mgr.modules {
+		var api resource.RPCSubtype
+		var ok bool
+		for a := range module.handles {
```
should we speed this up with some maps?

edaniels commented on module/modmanager/manager.go:
```diff
@@ -0,0 +1,370 @@
+	return depStrings
+}
+
+func (m *module) checkReady(ctx context.Context, addr string) error {
```
should we flip this to be that the client will call out to the server that it's ready?

Otterverse replied:
That was my original thought during early brainstorming, but... I believe this is simpler as is for a couple reasons:

1. The code flow is simpler if it can just wait for the module to start in the same function that's loading the module, rather than having to setup a channel to catch the incoming "ready" from another thread that's answering an incoming grpc request.
2. Currently, only one new proto was needed, and this represents the Module side, so no need for a second proto for the parent.
3. That also means less entanglement between the WebService and the module system. The "Lite" part of the WebService only handles resource services, same as the main web service does, so very little had to be changed there.
4. Lastly, this often means half as many sockets, as only modules that NEED a parent component ever connect to the parent. Having the module "check in" would mean there'd always be two connections per module.

edaniels replied:
okay fair points!

edaniels commented on module/modmanager/manager.go:
```diff
@@ -0,0 +1,370 @@
+	defer cancelFunc()
+	for {
+		req := &pb.ReadyRequest{ParentAddress: addr}
+		resp, err := m.client.Ready(ctxTimeout, req, grpc_retry.WithMax(5000))
```
5000 ms? if yes, can you make it a const somewhere

Otterverse replied:
No, it's "5000 attempts"... just an arbitrarily high number not meant to be hit (and this specific number was just kept from an example.) The real timeout is handled by the context timeout, but if WithMax isn't set (or is set to 0) the retry logic doesn't engage.

Added a comment to note the reason for the number.

edaniels commented on module/module.go:
```diff
@@ -0,0 +1,422 @@
+// NewHandlerMapFromProto converts protobuf to HandlerMap.
+func NewHandlerMapFromProto(ctx context.Context, pMap *pb.HandlerMap, conn *grpc.ClientConn) (HandlerMap, error) {
+	hMap := make(HandlerMap)
+	refClient := grpcreflect.NewClient(ctx, reflectpb.NewServerReflectionClient(conn))
```
we should get a nice looking error for when an SDK hasn't registered reflection services

Otterverse replied:
Added a "bail early" line in the loop just below this, so it fails out with a prettier message when that's the case.

edaniels commented on module/module.go:
```diff
@@ -0,0 +1,422 @@
+
+// NewModule returns the basic module framework/structure.
+func NewModule(address string, logger *zap.SugaredLogger) *Module {
+	m := &Module{
```
should do a logger.Named here

Otterverse replied:
Shouldn't it be left to the module author to pass in an already named logger? That feels more correct to me, so for now, updating the example module with logger.Named() there.

edaniels replied:
yep fine!

edaniels commented on module/module.go:
```diff
@@ -0,0 +1,422 @@
+	}
+
+	m.activeBackgroundWorkers.Add(1)
+	utils.PanicCapturingGo(func() {
```
should this be a ManagedGo?

Otterverse replied:
I don't believe so. If I understand correctly what MangedGo does, it would basically try to restart it if it panicked, and I don't think that's a good thing in this case, especially as this is only the grpc listner function. There's no promises that other states are correct, so it could quickly block on an unreleased mutex or something in the module's custom areas. I think the proper handling would likely be restarting the module itself (the parent would do this automatically) and as part of the future PR where reconfiguring an existing module is handled, reloading the config into a restarted module should be basically the same thing.

edaniels commented on module/module.go:
```diff
@@ -0,0 +1,422 @@
+		return err
+	}
+
+	oldMask := syscall.Umask(0o077)
```
if this alleviates my security concerns, can you write a detailed comment about the security assumptions here?

Otterverse replied:
Replied to two other security notes, and I think this is the third. Should hopefully make sense in aggregate, but I'll make a details statement here as requested.

The basic security assumption is that the socket (effectively a file) doesn't need to be any more secure than our config files (which contain the security keys.) E.g. as long as we set the permissions to only allow access to the owner (e.g. the user running the process) we can trust that the unix file security model is sufficient. Adding additional security here (e.g. encryption) is unneeded, since any attack that would work here would work to get the secrets from the config file as well (and thus allow decryption of everything anyway.)

tl;dr nothing to do right here I believe, but should add an ownership check before dialing (both in parent and child) to be extra cautious, and set the umask BEFORE directory creation in web.StartLite() (to prevent a race attack creating a false socket.)

edaniels commented on module/module.go:
```diff
@@ -0,0 +1,422 @@
+package module
```
given this is basically the SDK part of module, would it be better to call this the module client/child given we have the word parent? I want to make it obvious this file is for module writers

Otterverse replied:
I think that looks pretty clear once we have examples/tutorials. A module author will only ever need to see/use this package (not modmanager and others) so "import blah/blah/module" looks pretty intuitive to me, if I've set out to write a module. Least that's my gut feeling on the name.

edaniels replied:
alright

edaniels commented on module/modmanager/manager.go:
```diff
@@ -0,0 +1,370 @@
+		Args: []string{mgr.modules[cfg.Name].addr},
+		Log:  true,
+	}
+	mod.process = pexec.NewManagedProcess(pcfg, mgr.logger)
```
similar but better placed comment about" what testing/assurances do we have when a process dies?

Otterverse replied:
I'm making a TODO note about this topic, but I believe this fits in the next PR scope better. As mentioned in a previous comment, once "reconfigure" of a module works, reloading config into a crashed/restarted module should be trivial, and this can be shored up.

For now, this breaks pretty bad if the module itself dies. Basically a restart (of the parent) is required. But the same is true if the parent process dies, so from a practical standpoint, crashes that kill the process in a module are the same as crashing code in the parent.

edaniels replied:
sounds good

edaniels commented on robot/client/client.go:
```diff
@@ -593,9 +593,13 @@ func (rc *RobotClient) DiscoverComponents(ctx context.Context, qs []discovery.Qu
 		q := discovery.Query{
 			SubtypeName: resource.SubtypeName(disc.Query.Subtype),
-			Model:       disc.Query.Model,
+			Model:       m,
```
backwards compat q here

Otterverse replied:
I don't know enough about how/where this is used in App (elsewhere?) to provide a good answer. It passes tests here in RDK, but I expect changes will be needed on the App side. Also, I think I overlooked that this is using SubtypeName here instead of just "Subtype" so probably need to change that as well (like I did in most other "registry" like places)

edaniels commented on robot/impl/resource_manager.go:
```diff
@@ -335,9 +336,16 @@ func (manager *resourceManager) Close(ctx context.Context) error {
 			allErrs = multierr.Combine(allErrs, errors.Wrap(err, "error closing resource"))
 		}
+
+		if r.modules != nil && r.modules.IsModularResource(x) {
```
for these functions updates here, I'm wondering if there's any more succinct way to have this be built in to our "type system" such that you reconfigure a resource just the same as anything else. This may be pending the @cheukt @npmenard reconfig discussion.

edaniels commented on robot/web/web.go:
```diff
@@ -274,6 +292,64 @@ func RunWebWithConfig(ctx context.Context, r robot.LocalRobot, cfg *config.Confi
+		return errors.WithMessage(err, "module startup failed")
+	}
+	svc.liteAddr = dir + "/parent.sock"
+	oldMask := syscall.Umask(0o077)
```
security comment

Otterverse replied:
Saw your other comment about security on modmanager, and came here to double-check things. I need to move the umask up a few lines, before the mkdirtemp to ensure someone can't out-race the loading to create a fake module socket. Can add a check for ownership as well.

edaniels commented on robot/impl/local_robot.go:
```diff
@@ -130,14 +133,23 @@ func (r *localRobot) OperationManager() *operation.Manager {
+		web.Stop()
+	}
+	for s, svc := range r.internalServices {
+		if s == "web" {
```
use a const (hopefully already exists)

edaniels reviewed (changes requested):
Done with first pass! This is so cool. Lots of stuff here so you get lots of comments!

Overall Notes:
* Would like to remove IsMoving if possible
* Simplify registration internally (viam employees/contributors) and externally (module writer) - this would result in many less single line changes in this PR.
  *  Reducing usage of RegisterRPCLiteService
* Would like to have module server only offer what was explicitly registered for the purpose (i.e. not built ins)
* UDS security
* Module diff/reconfig
* How do implicit dependencies work?
* What happens if a module has a dependency that it itself offers? Is that all good? Does it end up talking to its dep via the parent or is it direct somehow?
* Tests :) (I know)
* Once PRs are in in other repos that this depends, let's do one last explicit round of review

Larger Discussion Point:
A module can support multiple models of components and services. There may be valid cases where someone wants to be able to run a board and motor that a module provides in two separate process spaces. The answer with this is that you leave it up to the user to do that; I think that's kind of hard. Instead, I think we should seriously consider limiting the server side of this for the first version such that a module can only provide one resource model and furthermore one instance of that model. I believe this would improve the DX for external users such that they build a module and `RegisterModule` once with a single model and constructor information. That means a module that wold provide 2 resources would need to be 2 modules. It also means the module manager would need to change slightly such that instead of having AddResource, it's a one time Start and then Configure for each instance of a component/service. Overall this should simplify DX and the architecture of this a bit. Curious what you and others think however.

edaniels commented:
@erh you should read over current comments and see how you feel about DX/UX here.

Otterverse commented on module/module.go:
```diff
@@ -0,0 +1,422 @@
+	return nil
+}
+
+func (m *Module) buildHandlerMap() {
```
But only if they explicitly import some built-in model. For the example module, the handlermap it builds is:

rdk:component:base -> acme:demo:mybase
rdk:component:gizmo -> acme:demo:mygizmo
rdk:service:navigation -> acme:demo:mynavigation
acme:service:summation -> acme:demo:mysum

Nothing else is being "accidentally" supported. But they could, obviously just `import _ go.viam.com/rdk/component/base/wheeled` and THEN, yes, it'd get automatically supported by the module. It'd still never get used though, as the registration lookup code in the parent should only go to a module IF the api/model isn't supported by something built-in.

Otterverse replied:
Explicit registration should now resolve this.

Otterverse commented on module/modmanager/manager.go:
```diff
@@ -0,0 +1,370 @@
+
+	for api, models := range mod.handles {
+		known := registry.ResourceSubtypeLookup(api.Subtype)
+		if known == nil {
```
This is to cover the case where the module is supporting a built-in subtype/api (or two different modules are supporting the same subtype.) We don't want to register a second time in that case, but it's normal/expected behavior.

Otterverse commented on components/arm/arm.go:
```diff
@@ -265,6 +272,12 @@ func (r *reconfigurableArm) Close(ctx context.Context) error {
 	return viamutils.TryClose(ctx, r.actual)
 }
 
+func (r *reconfigurableArm) IsMoving(ctx context.Context) (bool, error) {
```
It's needed, unfortunately. Without adding "IsMoving" into GRPC (and thus, the main, non Local Interfaces) the rpc.Client that represents modular resources within the parent can't provide the function, and thus the subtype service can't build status for any of those types, and then the UI can't display cards for them. The other option would be to remove "IsMoving" from all the status stuff.

I would REALLY like to get rid of the LocalXXXXX variations of interfaces as much as possible, and internally, if something absolutely needs some other function, then do a specific cast then and there for it.

edaniels replied:
But wouldn't that have been broken prior to this PR? Regarding the local interfaces, I think they may have been getting used incorrectly but they have a place for any kind of SDK, I believe

Otterverse replied:
So, prior to this, all resources in the resource manager were local objects WITH Local interfaces. So the "Status" function registered for the subtype services in question would just call IsMoving() (and other stuff.) Now any number of resources can be modular resources (which are grpc client objects) and thus the code breaks when trying to gather Status, because that client object couldn't support IsMoving().

Otterverse commented on components/arm/arm_test.go:
```diff
@@ -304,15 +304,6 @@ func TestReconfigurableArm(t *testing.T) {
 	test.That(t, err, test.ShouldBeNil)
 	test.That(t, reconfArm3, test.ShouldNotBeNil)
 
-	err = reconfArm1.Reconfigure(context.Background(), reconfArm3)
```
Sorry, yeah, this is non obvious, so should've explained.

With the change to IsMoving() there is no longer any difference between the Arm and LocalArm interfaces, so it's no longer expected to have errors here (and thus, can't test said errors.) Changing it to check for no error would also just duplicate already-run tests above.

Otterverse commented on components/arm/client.go:
```diff
@@ -126,3 +126,11 @@ func (c *client) GoToInputs(ctx context.Context, goal []referenceframe.Input) er
 	return generic.DoFromConnection(ctx, c.conn, c.name, cmd)
 }
+
+func (c *client) IsMoving(ctx context.Context) (bool, error) {
```
As noted, this is unfortunately needed to make things work for this PR. I do agree it would feel a lot more correct if split out, but it'd then have to be completed and merged first. Happy to discuss. Maybe another team can grab it go do it properly (and quickly) while I/we work through other stuff in this PR?

Otterverse commented on resource/model.go:
```diff
@@ -0,0 +1,140 @@
+const ModelFamilyDefaultName = ModelFamilyName("builtin")
+
+var (
+	// ModelFamilyDefault is the rdk:builtin model family for built-in resources.
```
If we had arms previously named "arm" and "armv2", yes. To clarify withactual examples: we have eva, ur5e, xarm6, etc. Those become "rdk:builtin:eva", "rdk:builtin:ur5e", "rdk:builtin:xarm6", etc.

Otterverse replied:
If we had arms previously named "arm" and "armv2", yes. To clarify withactual examples: we have eva, ur5e, xarm6, etc. Those become "rdk:builtin:eva", "rdk:builtin:ur5e", "rdk:builtin:xarm6", etc.

Otterverse commented on resource/model.go:
```diff
@@ -0,0 +1,140 @@
+// Validate ensures that important fields exist and are valid.
+func (f ModelFamily) Validate() error {
+	if f.Namespace == "" {
+		return errors.New("model namespace field for resource missing")
```
I followed what resource.Type and resource.Subtype do for validate, so that it remained consistent with that paradigm. Also, as there are no "paths" available here, that utility doesn't quite work. Wouldn't be opposed to handling this better in both places though... but as I went to look into that, I found the resource.*.Validate() methods are only ever entered one place, from the robotClient. They're never called outside of tests otherwise, at least as far as I can tell. Not sure what to do about that.

edaniels commented on resource/model.go:
```diff
@@ -0,0 +1,140 @@
+	}
+
+	var tempModel map[string]string
+	err := json.Unmarshal(data, &tempModel)
```
We had the linter but it kept finding false positives that couldn't be fixed :(. The reasoning is to reduce shadowing and variable introduction, although it's fairly low value I'll admit

Otterverse commented on config/reader.go:
```diff
@@ -75,12 +76,17 @@ var (
 // RegisterComponentAttributeMapConverter associates a component type and model with a way to convert all attributes.
-func RegisterComponentAttributeMapConverter(subtype resource.SubtypeName, model string, conv AttributeMapConverter, retType interface{}) {
+func RegisterComponentAttributeMapConverter(
+	subtype resource.Subtype,
```
If I understand you correctly, then yes. resource.SubtypeName is the last bit of the full identifier. E.g. just "arm" whereas the full subtype's string representation is "rdk:component:arm" and as the registry will now be handling custom resources, it needs to use the fully namespaced version as a key.

Otterverse commented on components/arm/fake/fake.go:
```diff
@@ -74,17 +75,17 @@ func NewArm(cfg config.Component, logger golog.Logger) (arm.LocalArm, error) {
 
 		switch converted.ArmModel {
-		case xarm.ModelName6DOF:
+		case xarm.ModelName6DOF.String():
```
Good catch, I think I misunderstand where the value was coming from. I assumed since it was from converted attributes, it'd using the JSON unmarshalling for Model where it'd automatically prepend "rdk:builtin" for configs with only the short model name. I think I see an easy fix though. `string(xarm.ModelName6DOF.Name)` and similar should do it (compare only against the short name)

Otterverse commented on examples/customresources/README.md:
```diff
@@ -0,0 +1,51 @@
+Standalone robot server (for use as a "remote") which serves serves a mygizmo, mynavigation, and mysum.
+
+### module
+The lighter weight (and preferred method) to integrate custom components. Serves all above APIs/components/services.
```
a top-level section of the config file called "modules" defines the name and path (to the binary) of a module, and that causes the module to get loaded during viam-server (parent process) startup. When it's loaded, it's then queried by the parent over grpc, and that query returns the APIs and models that the module supports. The modmanager bit in the parent then registers those components/services (and, if needed, new RPCSubtypes) just like built-in components, so the rest of the component/service loading happens as normal. Just the "Creator" functions registered redirect things back to the module(s) in question.

Otterverse commented:
Replying to @edaniels main comment in pieces here:

> Would like to remove IsMoving if possible

I believe it's 100% required for this PR to function, unless it's removed entirely from RDK. Changes _could_ be done independently, but would have to be approved and merged before this PR goes through.

> Simplify registration internally (viam employees/contributors) and externally (module writer) - this would result in many less single line changes in this PR.

I would like to see the same thing. I'm going to get some notes together on that, and we should have a brainstorming session on that if you're willing.

> Reducing usage of RegisterRPCLiteService

I believe this is part of the above.

> Would like to have module server only offer what was explicitly registered for the purpose (i.e. not built ins)

Not certain I understand the issue here. From my perspective, this is already exactly how this PR works. If a module author includes a "built-in" api (e.g. for Motor, because they want to support a custom motor) then it obviously uses it (and thus offers it) but the module author had to explicitly include rdk/motor for that to happen. It doesn't just blindly advertise every API or anything.

> UDS security (unix domain sockets)

Was addresses in more specific comments on the code and in slack. tl;dr for others, this should be resolved, and I made a couple of small changes (not pushed yet) to catch a security edge case. WIll explain with notes on those when merging.

> Module diff/reconfig

To clarify, reconfiguration of components/services IN a module works fine. The issue is only how to handle reconfiguration of the model itself. Right now, I think this is out of scope. Until we have versioning or similar, there's no "config" for a module beyond the name of its executable, so nothing to change. For now, a new module binary means a restart, just as a new viam-server binary itself requires a restart.

> How do implicit dependencies work?

It should be identical to how they work already. AFAIK, everything unique about "implicit" deps is handled well before the component creator function from the registry is called (and that passes the registry.Dependencies list.)

> What happens if a module has a dependency that it itself offers? Is that all good? Does it end up talking to its dep via the parent or is it direct somehow?

Currently it would go via the parent, but should work fine. I didn't write any shortcut code, as that'd be very language specific I think. Likewise, in many (most?) cases I expect a module author would shortcut things directly. E.g. a module supporting a base (with integrated motors) MIGHT choose to export both a Base and several Motors over grpc, but would likely not attempt to depend on them through the component system, but just directly access them in their local code. More often, they may only export the Base as a component.


> [The longer point about not supporting multiple models within a single module]

I believe that's a non-starter, and that this functionality is essential for many use cases, for three reasons:

1. A third-party may have multiple models in a product family that are very similar, and they don't want to maintain separate binaries for each of them. In our own codebase, we have a "xarm" component that supports the xarm5 and xarm6 models, with only minor differences between them. Thus, it'd make a lot of sense to be able to support both in the same module as well.

2. While our API model breaks things down to relatively fundamental components, actual products in the real world aren't always composed from the ground up like that. For example, the Trossen arms are functionally monolithic devices with a single serial input for commands, but in our ecosystem, they have to be represented by both an Arm AND a Gripper. If I had to define the gripper in a separate module/process from the arm, it'd be very difficult (if not impossible) to share the serial connection between the two. Similarly, many other composite devices exist.

3. In a similar fashion, many composite devices inherently ARE multiple instances of our resource types. A base with 4 integrated wheels (but only a single "brain" and line of communication) has to be able to support all four instances of a motor in one module. More concretely, the DMC4000 motor controller we have built in already is a single serial port device but supports up to 8 motor channels. The TMC5072 is a two-motor device as well.

While, to me, those three points alone make this essential functionality, I may not be understanding the full value in your reasoning. I agree the DX needs to be as simple as possible, and perhaps providing some SDK/helper functions would alleviate the concerns. For example, RDK's subtype.Service already helps handle multiple-instances of a resource, but we could probably do something better specifically for module use.

edaniels commented:
> Replying to @edaniels main comment in pieces here:
> 
> > Would like to remove IsMoving if possible
> 
> I believe it's 100% required for this PR to function, unless it's removed entirely from RDK. Changes _could_ be done independently, but would have to be approved and merged before this PR goes through.
> 
Okay fair!
> > Simplify registration internally (viam employees/contributors) and externally (module writer) - this would result in many less single line changes in this PR.
> 
> I would like to see the same thing. I'm going to get some notes together on that, and we should have a brainstorming session on that if you're willing.
> 
> > Reducing usage of RegisterRPCLiteService
> 
> I believe this is part of the above.
> 

Sounds good. Yeah my goal is to still have explicit module registration, not resource registration.

> > Would like to have module server only offer what was explicitly registered for the purpose (i.e. not built ins)
> 
> Not certain I understand the issue here. From my perspective, this is already exactly how this PR works. If a module author includes a "built-in" api (e.g. for Motor, because they want to support a custom motor) then it obviously uses it (and thus offers it) but the module author had to explicitly include rdk/motor for that to happen. It doesn't just blindly advertise every API or anything.
> 

See comment above. I don't think a blank import constitutes as explicit module registration. I want it to be a user intends to register their resource for modular purposes.

> > UDS security (unix domain sockets)
> 
> Was addresses in more specific comments on the code and in slack. tl;dr for others, this should be resolved, and I made a couple of small changes (not pushed yet) to catch a security edge case. WIll explain with notes on those when merging.
> 

Yep you addressed this.

> > Module diff/reconfig
> 
> To clarify, reconfiguration of components/services IN a module works fine. The issue is only how to handle reconfiguration of the model itself. Right now, I think this is out of scope. Until we have versioning or similar, there's no "config" for a module beyond the name of its executable, so nothing to change. For now, a new module binary means a restart, just as a new viam-server binary itself requires a restart.
> 

I'm asking about how to reconfigure when the `modules` section itself changes. What if you add a new module, is that processed? What if you delete one?

> > How do implicit dependencies work?
> 
> It should be identical to how they work already. AFAIK, everything unique about "implicit" deps is handled well before the component creator function from the registry is called (and that passes the registry.Dependencies list.)
> 

Implicit dependencies works by way of the configuration workflow though. I didn't see any proto methods that would cause a `Validate` to be called on a resource. Related to that, how can modules provide resource config validation prior to resource construction/instantiation?

> > What happens if a module has a dependency that it itself offers? Is that all good? Does it end up talking to its dep via the parent or is it direct somehow?
> 
> Currently it would go via the parent, but should work fine. I didn't write any shortcut code, as that'd be very language specific I think. Likewise, in many (most?) cases I expect a module author would shortcut things directly. E.g. a module supporting a base (with integrated motors) MIGHT choose to export both a Base and several Motors over grpc, but would likely not attempt to depend on them through the component system, but just directly access them in their local code. More often, they may only export the Base as a component.
> 
> > [The longer point about not supporting multiple models within a single module]
> 
> I believe that's a non-starter, and that this functionality is essential for many use cases, for three reasons:
> 
>     1. A third-party may have multiple models in a product family that are very similar, and they don't want to maintain separate binaries for each of them. In our own codebase, we have a "xarm" component that supports the xarm5 and xarm6 models, with only minor differences between them. Thus, it'd make a lot of sense to be able to support both in the same module as well.
> 

I don't think this point is essential necessarily.

>     2. While our API model breaks things down to relatively fundamental components, actual products in the real world aren't always composed from the ground up like that. For example, the Trossen arms are functionally monolithic devices with a single serial input for commands, but in our ecosystem, they have to be represented by both an Arm AND a Gripper. If I had to define the gripper in a separate module/process from the arm, it'd be very difficult (if not impossible) to share the serial connection between the two. Similarly, many other composite devices exist.
> 

**This** is essential I think and makes me retract my request.

>     3. In a similar fashion, many composite devices inherently ARE multiple instances of our resource types. A base with 4 integrated wheels (but only a single "brain" and line of communication) has to be able to support all four instances of a motor in one module. More concretely, the DMC4000 motor controller we have built in already is a single serial port device but supports up to 8 motor channels. The TMC5072 is a two-motor device as well.
> 

That's fair but in most cases I don't even know if a module in that scenario would need to expose anything other than the base itself.

> 
> While, to me, those three points alone make this essential functionality, I may not be understanding the full value in your reasoning. I agree the DX needs to be as simple as possible, and perhaps providing some SDK/helper functions would alleviate the concerns. For example, RDK's subtype.Service already helps handle multiple-instances of a resource, but we could probably do something better specifically for module use.

You offer a convincing argument. Maybe my main issue is the explicit registration bits!

erh commented on examples/customresources/module.json:
```diff
@@ -0,0 +1,72 @@
+	"modules": [
+		{
+			"name": "AcmeModule",
+			"path": "examples/customresources/module/module"
```
Can we make this a tad more descriptive?
path_binary? 
or something?

Otterverse replied:
Changing it to "executable_path" to follow the convention we have elsewhere (e.g. label_path, model_path, serial_path, etc.)

Otterverse commented:
Meeting with Cheuk and team yesterday, realized I misunderstood something about implicit dependencies. The "fallback" works normally, but there isn't currently a way to call a custom Validate() on configs for custom resources.

Otterverse commented on resource/resource.go:
```diff
@@ -157,6 +158,12 @@ func NewFromString(resourceName string) (Name, error) {
 
+// NewSubtypeFromString creates a new Subtype from string like: %s:%s:%s.
+func NewSubtypeFromString(subtypeName string) (Subtype, error) {
+	tmpName, err := NewFromString(subtypeName + "/ignored")
```
Okay, rewrote it to directly parse and not shortcut things with that /ignored trick.

Otterverse commented on config/module.go:
```diff
@@ -0,0 +1,34 @@
+	}
+
+	if m.Name == "parent" {
+		return errors.New("cannot use module name of 'parent'")
```
Made it a const.

As for a different directory... it just clutters things up. It seemed a lot cleaner to have only one temp directory with clearly labeled sockets inside it. This makes it easier during debugging to be able to know exactly what socket belongs to what (if when you want to directly grpcurl things or something.) Directory creation and cleanup only has to happen once as well. I don't think reserving one word for "parent" is unreasonable. Though if there's another word you'd prefer ("main" or something?) I wouldn't be opposed.

Otterverse commented on config/watcher.go:
```diff
@@ -132,26 +134,28 @@ func newFSWatcher(ctx context.Context, configPath string, logger golog.Logger) (
-						return
-					case configCh <- newConfig:
-					}
+					debounced(func() {
```
Explained in other comment I think so saw already, but tl;dr was "need to wait for ALL writes to happen so file is complete before reading"

Otterverse commented on examples/customresources/Makefile:
```diff
@@ -0,0 +1,20 @@
+
+buf:
+	cd ../../ && make tool-install
+	PATH="$(PATH):$(shell pwd)/../../bin/gotools/$(shell uname -s)-$(shell uname -m)" buf generate
```
Not easily. I can't include that file, because all those paths are relative, and rely on pwd, So it ends up setting a path of examples/customresources/bin/gotools (which doesn't exist.) Trying to rewrite the definition to work from BOTH locations results in something even uglier and more complex.

Otterverse commented on examples/customresources/README.md:
```diff
@@ -0,0 +1,51 @@
+* Start the server with `make run-module` Note: this automatically compiles the module itself first, which can be done manually with `make module`.
+* Run the client `make run-client`.
+
+## Reconfiguration
```
Addressed in the "big comments" of the PR, but tl;dr it's not handled. Without versioning (and auto downloads) any real change to the module itself would be something done manually, with a new module binary being updated/replace, and I don't think (for now) it's unreasonable to have that require a restart, the same as if the viam-server binary was updated.

Otterverse commented on robot/robot.go:
```diff
@@ -95,8 +96,20 @@ type LocalRobot interface {
+	WebAddress() (string, error)
+
+	// WebAddress returns the address of the web service.
+	LiteAddress() (string, error)
```
I'm fine with that. Want to have the bigger discussion about reworking the whole "RPCLite" registration first I think. See what still remains of this if that can be simplified.

Otterverse commented on examples/customresources/module.json:
```diff
@@ -0,0 +1,72 @@
+	"modules": [
+		{
+			"name": "AcmeModule",
+			"path": "examples/customresources/module/module"
```
Added documentation to the struct in config/module.go and updated the readme for the examples explaining as well.

Otterverse commented on module/modmanager/manager.go:
```diff
@@ -0,0 +1,370 @@
+	name    string
+	process pexec.ManagedProcess
+	handles modlib.HandlerMap
+	conn    *grpc.ClientConn
```
Once the socket is opened by both sides, they have a filehandle on it, and thus moving/renaming/deleting the actual "file" does nothing. This is another benefit of the socket model is not having to worry about traditional network problems/contingencies.

edaniels commented:
> Meeting with Cheuk and team yesterday, realized I misunderstood something about implicit dependencies. The "fallback" works normally, but there isn't currently a way to call a custom Validate() on configs for custom resources.

Cool. Let's do this in a later PR.

edaniels commented on robot/robot.go:
```diff
@@ -96,17 +96,17 @@ type LocalRobot interface {
-	// WebAddress returns the address of the web service.
-	LiteAddress() (string, error)
+	// ModuleAddress returns the address of the module service.
+	ModuleAddress() (string, error)
```
nit: ModuleServer or ModuleManager

edaniels commented on robot/web/web.go:
```diff
@@ -190,17 +190,17 @@ type Service interface {
 
-	// Start starts the web server
-	StartLite(context.Context) error
+	// StartModule starts the lightweight module service
```
should use ModuleManager or ModuleServer. Also we don't need to call it lightweight anymore here

edaniels commented on config/resource.go:
```diff
@@ -62,6 +62,7 @@ type Component struct {
 
 	Namespace     resource.Namespace           `json:"namespace"`
 	Type          resource.SubtypeName         `json:"type"`
+	API           resource.Subtype             `json:"api"`
```
this is probably fine but use `json"-"` with a comment that it's an experimental field

Otterverse commented on robot/robot.go:
```diff
@@ -95,8 +96,20 @@ type LocalRobot interface {
 	StartWeb(ctx context.Context, o weboptions.Options) error
 
+	// StartLite starts the local grpc lite server.
+	StartLite(ctx context.Context) error
```
I can (anad have) removed StartModule() from Robot, but ModuleAddress() is needed, otherwise webService() needs to become WebService() to expose ModuleAddress in turn. (It's consumed in the module manager.)

agreenb commented on services/datamanager/builtin/builtin.go:
```diff
@@ -63,8 +62,8 @@ const defaultCaptureBufferSize = 4096
-	Model              string               `json:"model"`
-	Type               resource.SubtypeName `json:"type"`
+	Model              resource.Model       `json:"model"`
+	Type               resource.Subtype     `json:"type"`
```
For quick background: we create a collector for each unique combination of [component name/model/type and method name/parameters]. 

* To create a collector on a component, we grab its nested data_manager config for the capture frequency and method name, and set the component name/model/type to that of the component config.
* To create a collector on a remote, we grab its nested data_manager config for the capture frequency, method name, and set the component name/type and remote robot name to that of the remote config.

This works well with modular resources by extracting the type into a string like `acme:demo:mygizmo`, but want to make sure there's parity between components and resources and is obvious what the user should input: e.g. should `Type` in ResourceLevelServiceConfig actually be named `API`?

Otterverse replied:
Actually, the more I look at this, I realize data_manager is the ONLY service using ResourceLevelServiceConfig so far. I'm going to undo this part of things for now, just go back to the SubtypeName as the "Type" field. In one or more of those later PRs, this should eventually get upgraded to triplets to support modular service_config services (it'll likely need to have a model as well.) I will notify when done so maybe you can take a final look.

agreenb replied:
Since we're already expecting a full triplet (e.g. rdk:component:arm/myRemoteArm) as the type for remotes and parse the name/type from it, this should be backwards compatible and work with Subtype as the "Type" field.

I believe the only thing I'd change here to not break existing data capture configs would be to set `attrs.Type = c.Type` since they won't have the API field below (`attrs.Type = c.API`).

agreenb replied:
And +1 to keeping ResourceLevelServiceConfig as-is — right now, the type is always just "data_manager" but that will change as we can now (woo!) easily implement different types of data managers.

Otterverse replied:
Regarding backward compatible... I can't directly use c.Type since it's only a subtypename, and c.API _should_ be filled out by any loaded config during the Validate() stage... but to be cautious, I'm changing it to `attrs.Type = c.ResourceName().Subtype` because the ResourceName() call will automatically generate the API field if it wasn't set, so should be safe in either case.

agreenb reviewed (commented):
This is really cool. Reviewed the data management and collectors code, and mostly just a high-level question about type vs api.

maximpertsov commented on robot/server/server.go:
```diff
@@ -165,6 +173,10 @@ func (s *Server) DiscoverComponents(ctx context.Context, req *pb.DiscoverCompone
+		if nonTriplet {
+			pbQuery.Subtype = string(discovery.Query.API.ResourceSubtype)
+			pbQuery.Model = string(discovery.Query.Model.Name)
+		}
```
cool, this should preserve discovery functionality with what app is currently sending

Otterverse commented:
Okay, I think this is getting pretty solid. Still need to write unit tests, but writing demos/examples helped expose a few things already. Notes since the last major review cycle:

1. I added a fallback for the Discovery service, and it appears to work normally with the current App when discovering a Camera, which is the only component currently using it. Had @maximpertsov  take a quick look, and he didn't find any other issues.
2. @agreenb  Can you do one final review of data_manager related things. I think it's all covered, but just please see if anything looks like it'll break in App if this merges.
3. @edaniels , please double-check the gRPC interceptors. I believe it's all working correctly now.
4. Changed config.Service.Type to be resource.Subtype instead of config.ServiceType. (We were just converting pointlessly in a lot of places, so just removed this extraneous type.)
5. examples/customresources is now less flat. Clearly seperated out three demos, the "complex module", a simple, one-file module, and the original remote server. Docs/Readmes have been updated as well.
6. Added to and rewrote a lot of docs and the readme.
7. And of course lots of other bug fixes and improvements/cleanup.

My outstanding TODO list on this is now down to writing tests, and opening JIRA issues for at least 4 significant follow-on/related things:

1. Support reconfigure/reloading of modules themselves. As part of that, handle restarting of crashed modules.
2. Pass validate() calls to module to deal with Implicit Deps generation.
3. Clean up reconfigure in RDK, make it simpler as a module too. Do resources get a "modular" flag maybe?
4. Dedupe config.Component/config.Service
  * Merge config.Component and config.Service, along with registry add/lookup functions for both.
  * Remove Namespace and Type fields in favor of API (triplet) field (that is a full resource.Subtype.)
  * ResourceLevelServiceConfig needs full API/Model (not just "type") to support third party services similar to data manager.
  * Discovery service needs support for triplets as well.

For tests, I have notes to myself (and some from eric) to focus on these areas:
* resource/model.go - all methods.
* resource/resource - Subtype and model, especially Json unmarshalling.
* config/diff_test.go - Test for model differences.
* config/resource_test.go - test defaults applied for model names as well as models that are fully specified
* config/proto_conversions.go - all subtype and model triplets and fallback.
* robot/server/server.go - Model triplet and discovery service.
* examples/customresources/* - use as much as possible.
* robot/impl/discovery_test.go - extend for supporting triplets


Now, with all that out of the way, please let me know if you have any of you have outstanding concerns, comments I might have overlooked or not answered, or things to add to the above lists.

Lastly, I know I'm submitting this minutes before the start of a 4 day weekend. Please don't let me interrupt your holidays though, so have a Happy Thanksgiving and we'll deal with this next week!

edaniels commented on examples/customresources/apis/summationapi/summationapi.go:
```diff
@@ -0,0 +1,181 @@
+
+// NewUnimplementedInterfaceError is used when there is a failed interface check.
+func NewUnimplementedInterfaceError(actual interface{}) error {
+	return utils.NewUnimplementedInterfaceError((Summation)(nil), actual)
```
looks like we use this once; should inline

edaniels commented on examples/customresources/apis/summationapi/summationapi.go:
```diff
@@ -0,0 +1,181 @@
+		RegisterRPCService: func(ctx context.Context, rpcServer rpc.Server, subtypeSvc subtype.Service) error {
+			return rpcServer.RegisterServiceServer(
+				ctx,
+				&pb.SummationService_ServiceDesc,
```
I think i mentioned it in previous year but we can address this in a follow up PR: internally, this amount of fields to popular feels okay, but externally, for module writers, this feels like a lot versus the necessary bits which are `NewServer`, `SummationService_ServiceDesc`, `RegisterSummationServiceHandlerFromEndpoint`, and RPCClient

edaniels commented on subtype/subtype.go:
```diff
@@ -69,3 +72,61 @@ func (s *subtypeSvc) Replace(r map[resource.Name]interface{}) error {
 	return nil
 }
+
+func (s *subtypeSvc) Add(n resource.Name, iface interface{}) error {
```
@npmenard bump

npmenard replied:
done just a couple of comments but other wise LGTM!

edaniels commented on module/module.go:
```diff
@@ -0,0 +1,424 @@
+	for _, c := range req.Dependencies {
+		name, err := resource.NewFromString(c)
+		if err != nil {
+			return &pb.AddResourceResponse{}, err
```
use nil, err

edaniels commented on module/module.go:
```diff
@@ -0,0 +1,424 @@
+	if err != nil {
+		return err
+	}
+	stat := info.Sys().(*syscall.Stat_t)
```
is there any more portable way to have us not cast to this type?

edaniels commented on module/server.go:
```diff
@@ -0,0 +1,97 @@
+}
+
+// ServeHTTP is unsupported.
+func (s *Server) ServeHTTP(http.ResponseWriter, *http.Request) {}
```
for any of these being unsupported, these should return an appropriate 5xx error

edaniels commented on robot/robot.go:
```diff
@@ -10,6 +10,7 @@ import (
 
 	"go.viam.com/rdk/config"
 	"go.viam.com/rdk/discovery"
+	modif "go.viam.com/rdk/module/modmanager/modmaninterface"
```
can this cycle avoidance get moved to `rdk/module/`?

Otterverse replied:
Moved to rdk/module/modmaninterface, so it's one less layer deep. Still has to be its own package though, (in case you meant trying to merge it into module itself.)

edaniels commented on robot/web/web.go:
```diff
@@ -845,8 +930,16 @@ func (svc *webService) initSubtypeServices(ctx context.Context) error {
-		if err := rs.RegisterRPCService(ctx, svc.rpcServer, subtypeSvc); err != nil {
-			return err
+
+		if rs.RegisterRPCService != nil && svc.rpcServer != nil && !mod {
```
fold into a larger block of `rs.RegisterRPCService != nil && svc.rpcServer != nil` with a nested mod conditional

edaniels commented on robot/server/server.go:
```diff
@@ -140,9 +141,24 @@ func (s *Server) ResourceRPCSubtypes(ctx context.Context, _ *pb.ResourceRPCSubty
 // component configurations.
 func (s *Server) DiscoverComponents(ctx context.Context, req *pb.DiscoverComponentsRequest) (*pb.DiscoverComponentsResponse, error) {
+	// nonTriplet indicates older syntax for type and model E.g. "camera" instead of "rdk:component:camera"
+	// TODO (@Otterverse) remove this logic when App and other dependencies are updated with triplets.
```
file a ticket for app

edaniels commented on module/modmanager/manager.go:
```diff
@@ -0,0 +1,300 @@
+
+	ctxTimeout, cancel := context.WithTimeout(ctx, time.Second*10)
+	defer cancel()
+	for {
```
might want to refactor this function into a few smaller functions to make testing and reading easier

edaniels commented on module/modmanager/manager.go:
```diff
@@ -0,0 +1,300 @@
+	return err
+}
+
+func (mgr *Manager) getModule(cfg config.Component) *module {
```
[minor] do not rely on nilness as a means of not found. use an explicit boolean param

edaniels commented on module/modmanager/manager.go:
```diff
@@ -0,0 +1,300 @@
+	for {
+		req := &pb.ReadyRequest{ParentAddress: addr}
+		// 5000 is an arbitrarily high number of attempts (context timeout should hit long before)
+		resp, err := m.client.Ready(ctxTimeout, req, grpc_retry.WithMax(5000))
```
this should use a backoff option too

edaniels commented on module/modmanager/manager.go:
```diff
@@ -0,0 +1,300 @@
+}
+
+// DepsToNames converts a dependency list to a simple string slice.
+func DepsToNames(deps registry.Dependencies) []string {
```
does this need to be exported?

edaniels reviewed (changes requested):
Done with this pass! This basically LGTM. Once my comments are addressed and tests are added, I'll do one final pass.

edaniels commented on module/server.go:
```diff
@@ -0,0 +1,97 @@
+// NewServer returns a new (module specific) rpc.Server.
+func NewServer(opManager *operation.Manager) rpc.Server {
+	s := &Server{server: googlegrpc.NewServer(
+		googlegrpc.UnaryInterceptor(opManager.UnaryServerInterceptor),
```
tests for this

Otterverse commented on registry/registry.go:
```diff
@@ -239,33 +244,38 @@ func DiscoveryFunctionLookup(q discovery.Query) (discovery.Discover, bool) {
-			validModels = append(validModels, splitName[1])
+			model, err := resource.NewModelFromString(splitName[1])
+			utils.UncheckedError(err)
+			validModels = append(validModels, model)
```
Doh!

Otterverse reviewed (commented):
Okay, resolved most of the smaller comments, and took more notes about where to focus testing. Working on that next (and the few larger comments not addressed.)

kkufieta commented on examples/customresources/README.md:
```diff
@@ -0,0 +1,81 @@
+This example demonstrates several ways RDK can be extended with custom resources. It contains several sections. Note that `make` is used throughout to help script commands. The actual commands being run should be printed as they are used. You can also look in the various "Makefile" named files throughout, to see the exact targets and what they do.
+
+## APIs
+APIs represent new types of components or services, with a new interface definition. They consist of protobuf descriptions for the wire level protocol, matching Go interfaces, and concrete Go implmenetations of a gRPC client and server.
```
implmenetations --> implementations ?

kkufieta commented on examples/customresources/README.md:
```diff
@@ -0,0 +1,81 @@
+Custom (service) api/protocol called "Summation" (acme:service:summation).
+
+### protos
+This folder contains the protobuf for the above two APIs. Only the .proto files are human modified. The rest is generated automatically by running "make" from within this directory. Note that the generation is performed using the "buf" command line tool, which itself is installed automatically as part of the make scripting. To generate protocols for other lanaguges, other tooling or commands may  ...
```
lanaguges -> languages ?

Otterverse commented:
Okay, this should be the final push. I merged main from an hour ago, all tests are passing, and I've added TODOs specifically linking to a variety of newly created JIRA issues, which are all collected in a master ticket here: https://viam.atlassian.net/browse/PRODUCT-337

Please go ahead and do any final reviews.

Also note this depends on small merges in API https://github.com/viamrobotics/api/pull/95 and goutils https://github.com/viamrobotics/goutils/pull/101 which will need to merge first.

kkufieta commented on module/doc.go:
```diff
@@ -0,0 +1,77 @@
+
+Under Golang, the module side of things tries to use as much of the "RDK" idioms as possible. Most notably, this includes the registry. So
+when creating modular components with this package, resources (and protocols) register their "Creator" methods and such during init() or
+during main(). They then are explicitly added via AddModelFromRegistry() so that merely importing a module doesn't add uneeded/unused
```
uneeded --> unneeded

kkufieta commented on examples/customresources/README.md:
```diff
@@ -0,0 +1,81 @@
+## Demos
+Each demo showcases an implementation of custom resources. They fall into two categories.
+
+* One is a module, which is the newer, preferred method of implementing a custom resource. It involves building a small binary that will (after configuration) be automatically started by a parent-viam server process, and communicate with it via gRPC over a Unix (file-like) socket on the local system. All configuration is done via the single parent robot, with relevant bits being passed to the mod ...
```
neccessary -> necessary

kkufieta commented on examples/customresources/README.md:
```diff
@@ -0,0 +1,81 @@
+
+* One is a module, which is the newer, preferred method of implementing a custom resource. It involves building a small binary that will (after configuration) be automatically started by a parent-viam server process, and communicate with it via gRPC over a Unix (file-like) socket on the local system. All configuration is done via the single parent robot, with relevant bits being passed to the mod ...
+
+* The other demo is the older, deprecated method, which is creating a standalone robot server (very similar to viam-server itself) with the new/custom component, and other non-needed parts stripped out, and then using that as a "remote" from a parent viam-server (even though it would technically run on the same local machine.) This requires two seperate configs, one for the parent, and one for th ...
```
seperate -> separate

kkufieta commented on examples/customresources/README.md:
```diff
@@ -0,0 +1,81 @@
+### complexmodule
+This demo is centered around a custom module that supports all four of the custom models above, including both custom APIs. A client demo is also included in a sub folder.
+
+#### Runnning
```
Runnning -> Running

kkufieta commented on examples/customresources/README.md:
```diff
@@ -0,0 +1,81 @@
+* Start the server `make run-module`
+  * This uses module.json
+  * This automatically compiles the module itself first, which can be done manually with `make module`.
+* In a seperate terminal, run the client with `make run-client` (or move into the client directory and simply run `make`.)
```
seperate -> separate

kkufieta commented on examples/customresources/README.md:
```diff
@@ -0,0 +1,81 @@
+* Start the server `make run-module`
+  * This uses module.json
+  * This automatically compiles the module itself first, which can be done manually with `make module`.
+* In a seperate terminal, run the client with `make run-client` (or move into the client directory and simply run `make`.)
```
seperate -> separate

kkufieta commented on examples/customresources/README.md:
```diff
@@ -0,0 +1,81 @@
+* Start the server `make run-module`
+  * This uses module.json
+  * This automatically compiles the module itself first, which can be done manually with `make module`.
+* In a seperate terminal, run the client with `make run-client` (or move into the client directory and simply run `make`.)
```
seperate -> separate

kkufieta commented on examples/customresources/README.md:
```diff
@@ -0,0 +1,81 @@
+  * This automatically compiles the module itself first, which can be done manually with `make module`.
+* In a seperate terminal, run the client with `make run-client` (or move into the client directory and simply run `make`.)
+
+#### Runnning
```
Runnning -> Running

edaniels commented on examples/customresources/demos/complexmodule/client/client.go:
```diff
@@ -0,0 +1,176 @@
+	defer robot.Close(context.Background())
+
+	logger.Info("---- Testing gizmo1 (gizmoapi) -----")
+	res, err := robot.ResourceByName(gizmoapi.Named("gizmo1"))
```
can you replace this with the idiom we have of gizmoapi.FromRobot

edaniels commented on examples/customresources/demos/complexmodule/module.go:
```diff
@@ -0,0 +1,66 @@
+	}
+	socketPath := args[1]
+
+	myMod, err := module.NewModule(ctx, socketPath, logger)
```
same comment here again

edaniels commented on examples/customresources/demos/remoteserver/server.go:
```diff
@@ -0,0 +1,50 @@
+
+// Arguments for the command.
+type Arguments struct {
+	ConfigFile                 string `flag:"config,usage=robot config file"`
```
formatting feels off here

edaniels commented on module/modmanager/manager.go:
```diff
@@ -0,0 +1,325 @@
+		return errors.WithMessage(err, "error while starting module "+mod.name)
+	}
+
+	if err := mod.dial(); err != nil {
```
I think if any of this fails, we need to kill the process via a defer

edaniels commented on module/modmaninterface/interface.go:
```diff
@@ -0,0 +1,23 @@
+	RemoveResource(ctx context.Context, name resource.Name) error
+	IsModularResource(name resource.Name) bool
+
+	NeedsModule(cfg config.Component) bool
```
name is a bit awkward. how about `Provides`?

edaniels commented on module/modmaninterface/interface.go:
```diff
@@ -0,0 +1,23 @@
+
+// ModuleManager abstracts the module manager interface.
+type ModuleManager interface {
+	AddModule(ctx context.Context, cfg config.Module) error
```
since this is all module based, how about just `Add`?

edaniels commented on module/modmaninterface/interface.go:
```diff
@@ -0,0 +1,23 @@
+	AddResource(ctx context.Context, cfg config.Component, deps []string) (interface{}, error)
+	ReconfigureResource(ctx context.Context, cfg config.Component, deps []string) error
+	RemoveResource(ctx context.Context, name resource.Name) error
+	IsModularResource(name resource.Name) bool
```
It feels like as a user `IsModularResource` and `NeedsModule` are basically the same. Can we just merge them and expect a resource.Name (or better Type) to always be passed?

edaniels commented on resource/resource.go:
```diff
@@ -110,6 +132,29 @@ func (s Subtype) String() string {
 	return fmt.Sprintf("%s:%s", s.Type, s.ResourceSubtype)
 }
 
+// UnmarshalJSON pareses namespace:type:subtype strings to the full Subtype{} struct.
```
Subtype{} -> Subtype

cheukt replied:
pareses -> parses

edaniels commented on resource/resource.go:
```diff
@@ -91,6 +100,16 @@ func NewSubtype(namespace Namespace, rType TypeName, subtype SubtypeName) Subtyp
 }
 
+// NewDefaultSubtype creates a new Subtype based on parameters passed in.
+func NewDefaultSubtype(subtype SubtypeName, isSvc bool) Subtype {
```
instead of bool why not just ResourceType

edaniels commented on registry/registry.go:
```diff
@@ -203,6 +224,13 @@ func RegisterResourceSubtype(subtype resource.Subtype, creator ResourceSubtype)
 }
 
+// DeregisterResourceSubtype removes a previously registered subtype.
+func DeregisterResourceSubtype(subtype resource.Subtype) {
```
Nice english

edaniels commented on operation/opid.go:
```diff
@@ -134,6 +135,11 @@ func (m *Manager) createWithID(ctx context.Context, id uuid.UUID, method string,
 
+	o := m.Find(id)
+	if o != nil {
+		m.logger.Warnw("attempt to create duplicate operation", "id", id.String(), "method", method)
```
ty

edaniels commented on module/modmanager/manager_test.go:
```diff
@@ -0,0 +1,159 @@
+			Name:    "simple-module",
+			ExePath: modExe,
+		}
+		err := mgr.AddModule(ctx, modCfg)
```
I'd be weary of making subtests that mutate one variable that gets used in the next. I'd suggest making a setup and teardown function that runs per sub test

edaniels commented on module/module.go:
```diff
@@ -0,0 +1,433 @@
+	if creator.ReflectRPCServiceDesc == nil {
+		m.logger.Errorf("rpc subtype %s doesn't contain a valid ReflectRPCServiceDesc", api)
+	}
+	rpcST := resource.RPCSubtype{
```
prefer a longer name like rpcSubtype. it's easier to read than guess IMO

edaniels commented on module/module_interceptors_test.go:
```diff
@@ -0,0 +1,182 @@
+		}},
+	}
+
+	// cfg.Modules[0].ExePath = utils.ResolveFile("module/testmodule/run.sh")
```
remove?

edaniels commented on module/module_interceptors_test.go:
```diff
@@ -0,0 +1,182 @@
+	return rc, gc, conn, nil
+}
+
+func mkCfg() (string, string, error) {
```
more verbose the better. makeConfig

edaniels commented on robot/impl/local_robot.go:
```diff
@@ -131,19 +134,28 @@ func (r *localRobot) OperationManager() *operation.Manager {
-			return err
+	web, err := r.webService()
+	if err == nil {
+		web.Stop()
```
so we Stop and Close web?

edaniels commented on robot/impl/local_robot.go:
```diff
@@ -218,6 +235,24 @@ func (r *localRobot) StopWeb() error {
+	return webSvc.Address(), nil
+}
+
+// LiteAddress return the lite grpc service's address.
```
lite -> module

edaniels commented on module/testmodule/main.go:
```diff
@@ -0,0 +1,89 @@
+	}
+
+	switch req["command"] {
+	case "sleep":
```
neat idea

edaniels commented on resource/model.go:
```diff
@@ -0,0 +1,151 @@
+		return err
+	}
+	if m.Name == "" {
+		return errors.New("model name field for resource missing")
```
isn't it just name missing?

edaniels reviewed (approved):
This LGTM. There are plenty of nits though in this review and one minor thing that I'd like addressed prior to merging. That being said, the overall production code is LGTM. Great work James.

Otterverse commented on components/servo/server.go:
```diff
@@ -66,3 +66,16 @@ func (server *subtypeServer) Stop(ctx context.Context, req *pb.StopRequest) (*pb
+
+// IsMoving queries of a component is in motion.
+func (server *subtypeServer) IsMoving(ctx context.Context, req *pb.IsMovingRequest) (*pb.IsMovingResponse, error) {
+	base, err := server.getServo(req.GetName())
```
Doh! Yeah, did that in several components. Fixed now.

cheukt commented on config/config.go:
```diff
@@ -63,6 +64,12 @@ func (c *Config) Ensure(fromCloud bool) error {
 
+	for idx := 0; idx < len(c.Modules); idx++ {
+		if err := c.Modules[idx].Validate(fmt.Sprintf("%s.%d", "modules", idx)); err != nil {
+			return err
```
since we allow partial start up now, should we error out just because a module config is wrong?

npmenard commented on subtype/subtype.go:
```diff
@@ -45,27 +48,93 @@ func (s *subtypeSvc) Resource(name string) interface{} {
+
+	s.resources[name] = iface
+	shortcut := getShortcutName(name)
+	if shortcut != name {
```
I think we should remove this check, if both a local resource "NameA" and a remote resource "remote1:NameA" exists the shortcut shouldn't be registered

npmenard replied:
looks good thanks :)

npmenard commented on subtype/subtype.go:
```diff
@@ -45,27 +48,93 @@ func (s *subtypeSvc) Resource(name string) interface{} {
+	_, ok := s.resources[name]
+	if ok {
+		delete(s.resources, name)
+	}
```
nit: return an error if the resource doesn't exist unless we want to silently allow multiple deletions of the same resource

npmenard commented on subtype/subtype.go:
```diff
@@ -45,27 +48,93 @@ func (s *subtypeSvc) Resource(name string) interface{} {
+	}
+
+	// case: remote1:nameA and remote2:nameA both existed, and remote2:nameA is being deleted, restore shortcut to remote1:nameA
+	for k := range s.resources {
```
If we expected this function to be often called I would only delete the entry if it isn't an empty string and otherwise proceed with attempting to find a shortcut. So that we don't over abuse the GC

npmenard commented on subtype/subtype.go:
```diff
@@ -45,27 +48,93 @@ func (s *subtypeSvc) Resource(name string) interface{} {
+
+	// case: remote1:nameA and remote2:nameA both existed, and remote2:nameA is being deleted, restore shortcut to remote1:nameA
+	for k := range s.resources {
+		if shortcut == getShortcutName(k) && name != getShortcutName(k) {
```
same comment as above

Otterverse reviewed (commented):
I believe I've addressed everything from this round of reviews, except for two pending questions. One to @npmenard about subtype, as I want confirmation the tests I added there cover the case he's concerned about (and therefore it wasn't a problem.) Second, @edaniels I have questions on how you want me to approach the web_test.go comment about opID testing on streams.

Otherwise, hopefully all good. I'll get to the api/goutils comments in the morning.

edaniels commented on resource/resource.go:
```diff
@@ -1,8 +1,24 @@
+and useful mostly for organization/grouping. Note that each "tier" contains the tier to the left it. Such that ModelFamily contains
+Namespace, and Model itself contains ModelFamily.
+
+An example resource (say, a motor) may use the motor API, and thus have the Subtype "rdk:component:motor" and have a model such as
```
nice

edaniels commented on robot/web/web_test.go:
```diff
@@ -863,13 +927,13 @@ func TestRawClientOperation(t *testing.T) {
 	md, err := streamClient.Header()
 	test.That(t, err, test.ShouldBeNil)
-	checkOpID(md)
+	checkOpID(md, false)
```
ah yes, please use my little thing in that case. i'd just copy it over since it's not a real resource we work with. it's okay to be filtering stream status. the test when i wrote it just looked for something that streamed :)

edaniels commented:
left remaining comments! thanks for that quick turnaround time
